ui/src/main/java/com/kotcrab/vis/ui/layout/FlowGroup.java

```diff
@@ -0,0 +1,402 @@
+package com.kotcrab.vis.ui.layout;
+
+import com.badlogic.gdx.scenes.scene2d.Actor;
+import com.badlogic.gdx.scenes.scene2d.Touchable;
+import com.badlogic.gdx.scenes.scene2d.ui.WidgetGroup;
+import com.badlogic.gdx.scenes.scene2d.utils.Layout;
+import com.badlogic.gdx.utils.SnapshotArray;
+
+/**
+ * Arranges actors to flow in a specified layout direction using up available space and, if sensible, expanding in that direction.
+ * <br/>
+ * For horizontal layout direction ({@code horizontal=true}), attempts to expand to the desired width, creates new rows and expands vertically as necessary.
+ * Children automatically overflow to the next row when necessary.
+ * <br/>
+ * For vertical layout direction ({@code horizontal=false}), attempts to expand to the desired height, creates new columns and expands horizontally as necessary.
+ * Children automatically overflow to the next column when necessary.
+ * <br/>
+ * <br/>
+ * Can be embedded in a scroll pane. To ensure proper flowing (instead of expanding in the layout direction), scrolling in the layout direction should be disabled, i.e., horizontal layout direction should have scrolling in x direction disabled and vertical layout direction should have scrolling in y direction disabled.
+ * <br/>
+ * <br/>
+ * This is a more versatile implementation of a flow group subsuming
+ * {@link com.kotcrab.vis.ui.layout.HorizontalFlowGroup HorizontalFlowGroup} and {@link com.kotcrab.vis.ui.layout.VerticalFlowGroup VerticalFlowGroup}.
+ * <br/>
+ * <br/>
+ * Key differences:
+ * <ul>
+ *   <li>Can be configured to have a layout direction as either horizontal ({@code horizontal=true}) or vertical ({@code horizontal=false}).</li>
+ *   <li>Layout direction can be changed programmatically during runtime.</li>
+ *   <li>If available, uses up all necessary space in the layout direction, i.e., when {@code horizontal=true}, attempts to expand horizontally and, when {@code horizontal=false}, attempts to expand vertically (instead of using the specified width/height as before).</li>
+ *   <li>Adds spacing only between children, but not after the last element.</li>
+ *   <li>When even the first child does not fit its row/column, space is no longer placed before it.</li>
+ * </ul>
+ *
+ * @author ccmb2r
+ * @since 1.4.7
+ */
+public class FlowGroup extends WidgetGroup {
+	private static final float DEFAULT_SPACING = 0;
+
+	private boolean horizontal;
+	private float spacing;
+
+	//Internal
```

Comment:
Unnecessary comment

Reply:
Hey kotcrab,

lost track of this for a bit, still interested in contributing though!

Regarding your questions/comments:

**1a. Will the resizing logic really work?**

Yes, I'd say it will. What it does is take one attempt at resizing and either use the extended size (e.g., if within a ScrollPane that allows scrolling in the layout direction and, thus, can accommodate more space) or live with the existing size as the calculation will yield the same result then.


**1b. Is this done somewhere in libGDX?**

Yes, the [HorizontalGroup uses a similar procedure in its layoutWrapped () method](https://github.com/libgdx/libgdx/blob/8539c3baf6a901f5272666d7305cb1234cbb27f3/gdx/src/com/badlogic/gdx/scenes/scene2d/ui/HorizontalGroup.java#L209) to determine at which width to wrap. 


**2. Unnecessary comment.**

I'll remove it.


**3. Further integration**

I was not sure how invasive contributions may be but would be happy to integrate this further into VisUI. Adapting the test would certainly make sense if you are happy with, perspectively, using FlowGroup as a replacement for the current layouts. Here are a few further thoughts on how to integrate the new changes while trying to maintain backward compatibility:


**a) FlowGroup as parent of HorizontalFlowGroup and VerticalFlowGroup**

One could make FlowGroup the parent of HorizontalFlowGroup and VerticalFlowGroup getting rid of their specific implementations and, perspectively, removing the specialized classes altogether. The behavior is very similar so that there would only be minor edge cases where it would affect existing applications, namely that FlowGroup does not add excess space in unfavorable resizing scenarios and that it uses up available space in the layout direction by expanding if possible. FlowGroup behaves similarly enough in a ScrollPane I'd say.

One caveat though: Both HorizontalFlowGroup and VerticalFlowGroup would inherit the methods [isHorizontal ()](https://github.com/kotcrab/vis-ui/blob/cf8c46acb6589443ba2d1367977e3a11d8570176/ui/src/main/java/com/kotcrab/vis/ui/layout/FlowGroup.java#L66) and [setHorizontal (boolean)](https://github.com/kotcrab/vis-ui/blob/cf8c46acb6589443ba2d1367977e3a11d8570176/ui/src/main/java/com/kotcrab/vis/ui/layout/FlowGroup.java#L70), which do not make sense in their context. As a workaround, one could overwrite these methods and have them throw an UnsupportedOperationException with a pointer to the new FlowGroup class.

In this particular use case, one could leave the current classes in DragPane and TabbedPane until, ultimately, HorizontalFlowGroup and VerticalFlowGroup are removed from VisUI (at which point, FlowGroup would be used in their place).


**b) Replacing HorizontalFlowGroup and VerticalFlowGroup in existing VisUI code**

It would certainly be possible to replace the use of HorizontalFlowGroup and VerticalFlowGroup in existing VisUI code, i.e., in DragPane and TabbedPane (which seem to be the only internal uses apart from tests). The caveat might be if existing users depend on these particular classes in their extensions. However, it seems that the concrete classes are mostly encapsulated in existing controls. TabbedPane should be relatively easy to adapt without effects on existing users. DragPane still has some deprecated methods that expose the concrete implementation classes HorizontalFlowGroup and VerticalFlowGroup as well as their superseding methods that determine whether the DragPane uses a horizontal or vertical flow. I could certainly adapt the latter but keeping the prior would pose a problem unless the solution under a) was used as a temporary workaround.

I'd be happy to hear your thoughts on this and proceed accordingly.